# Genome Sequencing Bioinformatics - Africa GSBAfrica
Base for the Genome Sequencing Bioinformaticss course repository and manual

## Course overview
In collaboration with [H3ABioNet](https://h3abionet.org/), we are pleased to announce the Genome Sequencing Bioinformatics Africa course. This is a new iteration of the previous course – “[Next Generation Sequencing Bioinformatics – Africa 2022](https://coursesandconferences.wellcomeconnectingscience.org/event/next-generation-sequencing-bioinformatics-africa-20220322/)” 

High throughput sequencing is an essential tool in genetic and genomic analysis. It is increasingly important for experimental scientists to gain the bioinformatics skills required to analyse the large volumes of data produced by sequencers. This course will equip participants with the essential informatics skills required to begin analysing sequencing data and apply some of the most commonly used tools and resources for sequence data analysis. 

The programme will cover prominent sequencing technologies, algorithmic theory and principles of bioinformatics, with a strong focus on practical computational sessions using sequence analysis techniques and tools applicable to any species or genome size. A variety of applications will be covered from post-sequencing analysis, quality control, alignment, assembly, and variant calling.

This course will apply a blended learning format consisting of locally coordinated classrooms referred to as “distributed classrooms”. See the publications by [Gurwitz et al.](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005715) and [Ras et al](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008640) for more information. The local classrooms for this course may run virtually using Zoom.

**Time commitment:**

Contact sessions will run on Tuesdays and Thursdays lasting for 4 hours per session.

**Target audience:**

The course is aimed at postdoctoral scientists, senior PhD students, junior faculty members or clinicians/healthcare professionals based in Africa who are actively engaged in or soon to commence research involving  sequencing data analysis.

**Please note:**

The practical sessions will be taught exclusively through Unix/Linux. Therefore, participants are required to have some previous experience using the Linux operating system. This will be essential for participants to fully benefit from the course. There are numerous online introductory tutorials to the UNIX/Linux operating system and command line, including:
- http://www.ee.surrey.ac.uk/Teaching/Unix 
- http://swcarpentry.github.io/shell-novice/  



[Course website](https://coursesandconferences.wellcomeconnectingscience.org/event/genome-sequencing-bioinformatics-africa-20230815/)

## Instructors
- [Instructor](link),Affiliation

### Overview


**The programme will cover the following core topics:**

- Intro to Unix/Linux & running workflows
- Introduction to Sequencing Technologies
- Sequencing data pre-processing and QC
- Alignment to reference sequences
- Variant calling and annotation
- Learning outcomes

**On completion of the course, participants should expect to be able to:**

- Use the unix command-line as a tool for data analysis
- Describe the different sequencing data file formats available
- Perform QC assessment of high throughput sequencing data
- Explain the algorithmic concepts behind read alignment, variant calling and structural variant detection
- Perform read alignment, variant calling and structural variation detection using standard tools

### Detailed timetable
[View Timetable here "in dev"]()



## Course manual
**Virtual Machine Download and Installation**        

[Download the Virtual Machine with Globus](modules/Download_course_Virtual_Machine.md)        

[Install the Virtual Machine with VirtualBox](modules/VM_Installationwith_Virtualbox.md)         

[Issues and Fixes](modules/VM_Issues_and_Fixes.md)        

[VM Frequently asked Questions FAQs](modules/VM_Question_and_Answer.md)        



**Module 1 - Intro to Unix/Linux**   
[Online Manual - Intro to Unix/Linux Session 1 ](modules/Module_1_Session1_practical_assignment.docx.md)

[Online Manual - Intro to Unix/Linux Session 2 ](modules/Module_1_Session2_practical_assignment.docx.md)


**Module 2 - Introduction to Sequencing Technologies**    
[Online Manual - Introduction to Sequencing Technologies](modules/Module2_Session1_base.md)

**Module 3 - Sequencing data formats and QC**  

[Online Manual - Sequencing data formats and QC Session 1](modules/Module3_Day1_Assignment.docx.md)

[Online Manual - Sequencing data formats and QC Session 2](modules/Module3_Day2_Data_formats_Assignment.md)

  
**Module 4 - Alignment to Reference**      

[PDF Manual - Alignment to Reference](modules/read-alignment.practicalexercises.20210504.pdf)    

**Module 5 - Variant Calling - Human**  

[Base Instructions - Variant Calling - Human](modules/Module5/Module5_Session1_base.md) 

[Online Manual - Human Variant Calling](modules/Module5/Module5_Session1_practical_assignment.md) 

[PDF Manual - Human Variant Calling ](modules/Module5/Module5_Session1_practical_assignment.pdf) 


**Module 6 - Variant Calling - Pathogen**  
Will be released 19 Sep

<!--- [Online Manual - Variant Calling - Pathogen](modules/) --->


**Appendix**      
  


Any reuse of the course materials, data or code is encouraged with due acknowledgement.

******
## License
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0)</a>.

